mava/environment_loop.py

```diff
@@ -481,10 +494,46 @@ def should_terminate(episode_count: int, step_count: int) -> bool:
                 num_steps is not None and step_count >= num_steps
             )
 
+        def should_run_loop(eval_condtion: Tuple) -> bool:
+            """Check if the eval loop should run in current step.
+
+            Args:
+                eval_condtion : tuple containing interval key and count.
+
+            Returns:
+                a bool indicatings if eval should run.
+            """
+            should_run_loop = False
+            eval_interval_key, eval_interval_count = eval_condtion
+            counts = self.get_counts()
+            if counts:
+                count = counts.get(eval_interval_key)
+                # We run eval loops around every eval_interval_count (not exactly every
+                # eval_interval_count due to latency in getting updated counts).
+                should_run_loop = (
+                    (count - self._last_evaluator_run_t) / eval_interval_count
+                ) >= 1.0
+                if should_run_loop:
+                    self._last_evaluator_run_t = int(count)
+            return should_run_loop
+
         episode_count, step_count = 0, 0
+
+        # Currently, we only use intervals for eval loops.
+        environment_loop_schedule = (
+            self._executor._evaluator and self._executor._interval
+        )
+        if environment_loop_schedule:
+            eval_condtion = check_count_condition(self._executor._interval)
+
         while not should_terminate(episode_count, step_count):
-            result = self.run_episode()
-            episode_count += 1
-            step_count += result["episode_length"]
-            # Log the given results.
-            self._logger.write(result)
+            if (not environment_loop_schedule) or should_run_loop(eval_condtion):
+                result = self.run_episode()
+                episode_count += 1
+                step_count += result["episode_length"]
+                # Log the given results.
+                self._logger.write(result)
+
+            # We need to get the latest counts if we are using eval intervals.
+            if environment_loop_schedule:
+                self._executor.update()
```

Comment:
Is this update not already performed in the run_episode call?

Reply:
So this update is important to get the latest counts.

Since these loops run on different processes, while the evaluator is waiting, it doesn't run the loop and so it vars counts don't get updated (this part doesn't [run](https://github.com/instadeepai/Mava/blob/52cd450c08e695380b81023f45228fe7a4e43749/mava/environment_loop.py#L439)). 

So while the evaluator is waiting, it still need to call `self._executor.update()` to get latest counts from the executor runs. 

Not sure if that makes sense?

Reply:
I see. That makes sense yes. Should there not maybe be a time.sleep somewhere in here? This is so that the evaluator does not constantly speak to the variable_server while waiting. And also `self._executor.update()` only updates the executor every 1000 steps right? So we maybe need to change that setting to 1 in the case of using an evaluator that has waiting intervals?